src/python/pants/engine/isolated_process.py

```diff
@@ -2,101 +2,86 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 import logging
+from dataclasses import dataclass
+from typing import Dict, Optional, Tuple, Union
 
 from pants.engine.fs import EMPTY_DIRECTORY_DIGEST, Digest
 from pants.engine.platform import PlatformConstraint
 from pants.engine.rules import RootRule, rule
-from pants.util.objects import (
-  Exactly,
-  TypedCollection,
-  datatype,
-  hashable_string_list,
-  string_optional,
-)
 
 
 logger = logging.getLogger(__name__)
 
 _default_timeout_seconds = 15 * 60
 
 
-class ProductDescription(datatype([('value', str)])): pass
+@dataclass(frozen=True)
+class ProductDescription:
+  value: str
 
 
-class ExecuteProcessRequest(datatype([
-  ('argv', hashable_string_list),
-  ('input_files', Digest),
-  ('description', str),
-  ('env', hashable_string_list),
-  ('output_files', hashable_string_list),
-  ('output_directories', hashable_string_list),
-  # NB: timeout_seconds covers the whole remote operation including queuing and setup.
-  ('timeout_seconds', Exactly(float, int)),
-  ('unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule', Digest),
-  ('jdk_home', string_optional),
-])):
+@dataclass(unsafe_hash=True)
+class ExecuteProcessRequest:
   """Request for execution with args and snapshots to extract."""
 
   # TODO: add a method to hack together a `process_executor` invocation command line which
   # reproduces this process execution request to make debugging remote executions effortless!
-  def __new__(
-    cls,
-    argv,
-    input_files,
-    description,
-    env=None,
-    output_files=(),
-    output_directories=(),
-    timeout_seconds=_default_timeout_seconds,
-    unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule=EMPTY_DIRECTORY_DIGEST,
-    jdk_home=None,
-  ):
-    if env is None:
-      env = ()
-    else:
-      if not isinstance(env, dict):
-        raise cls.make_type_error(
-          "arg 'env' was invalid: value {} (with type {}) must be a dict".format(env, type(env)))
-      env = tuple(item for pair in env.items() for item in pair)
-
-    return super().__new__(
-      cls,
-      argv=argv,
-      env=env,
-      input_files=input_files,
-      description=description,
-      output_files=output_files,
-      output_directories=output_directories,
-      timeout_seconds=timeout_seconds,
-      unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule=unsafe_local_only_files_because_we_favor_speed_over_correctness_for_this_rule,
-      jdk_home=jdk_home,
-    )
-
-
-class MultiPlatformExecuteProcessRequest(datatype([
-  ('platform_constraints', hashable_string_list),
-  ('execute_process_requests', TypedCollection(Exactly(ExecuteProcessRequest))),
-])):
-  # args collects a set of tuples representing platform constraints mapped to a req, just like a dict constructor can.
-
-  def __new__(cls, request_dict):
+  argv: Tuple[str, ...]
```

Comment:
There's going to be a comment added in the README about this in #8181!

Reply:
Yes, we must use `Tuple[T, ...]` for immutability. I'm going to hold off on a comment because this is a pretty standard pattern we use and Danny's PR will document that in the README.